### PR TITLE
protect s3tokens and ec2tokens with lifesaver middleware

### PR DESCRIPTION
**Pull request by bbobrov**

### CODE REVIEW

File: cc/keystone/middleware/lifesaver.py [hunk 1]
```diff
@@ -81,6 +81,23 @@ def get_user(self, request):
...
                                 user = 'ac-' + user
 
+            elif '/v3/s3tokens' == request.path and 'POST' == request.method:
+                # s3tokens never contains user id, so use credential's `access`
+                # field to identify it
+                body = request.json_body
+                credentials = (
+                    body.get('credentials') or
+                    body.get('credentials') or
+                    body.get('ec2Credentials')
```
Comment (stanislav-zaprudskiy) [inline cc/keystone/middleware/lifesaver.py]:
```suggestion
                    body.get('credentials') or
                    body.get('ec2Credentials')
```

**Review by stanislav-zaprudskiy — CHANGES_REQUESTED** [review]

File: cc/keystone/middleware/lifesaver.py (hunk 1 above)
Comment (bbobrov) [reply cc/keystone/middleware/lifesaver.py]: there is indeed a mistake here

Comment (bbobrov) [reply cc/keystone/middleware/lifesaver.py]: https://github.com/openstack/keystone/blob/master/keystone/api/_shared/EC2_S3_Resource.py#L80-L93

**Review by stanislav-zaprudskiy — APPROVED** [review]

**Comment by bbobrov** [comment]
I have tested this in qa-de-3 and it works. However, it might be good to cover /ec2tokens API with this code as well, because it might also flood us. Let me do it...

**Review by stanislav-zaprudskiy — APPROVED** [review]
🚀

**Pull request merged**